src/gensxg.c

```diff
@@ -0,0 +1,377 @@
+    exit(1);
+  }
+  return (uint64_t)mktime(&date);
+}
+
+Options parse_options(int argc, char* const argv[]) {
+  Options result;
+  memset(&result, 0, sizeof(result));
+  // Set default values.
+  result.help = false;
+  result.content_type = "text/html";
+  time_t now = time(NULL);
+  result.date = (uint64_t)now;
+  result.mi_record_size = 4096;
+
+  int opt;
+  int longindex;
+  while ((opt = getopt_long_only(argc,
+                                 argv,
+                                 "h:u:r:v:e:p:k:c:t:d:x:m:",
+                                 kOptions,
+                                 &longindex)) != -1) {
+    switch (opt) {
+      case 'h':
+        result.help = true;
+        return result;
+      case 'u':
+        result.url = optarg;
+        break;
+      case 'r':
+        result.cert_url = optarg;
+        break;
+      case 'v':
+        result.validity_url = optarg;
+        break;
+      case 'e':
+        result.certificate = optarg;
+        break;
+      case 'p':
+        result.public_key = optarg;
+        break;
+      case 'k':
+        result.private_key = optarg;
+        break;
+      case 'c':
+        result.content = optarg;
+        break;
+      case 't':
+        result.content_type = optarg;
+        break;
+      case 'd':
+        result.date = parse_time(optarg);
+        break;
+      case 'x':
+        result.expire = parse_time(optarg);
+        break;
+      case 'm':
+        result.mi_record_size = strtoull(optarg, NULL, 0);
+        break;
+      default:
+        exit(1);
+    }
+  }
+
+  if (result.expire == 0) {
+    result.expire = result.date + 60 * 60 * 24 * 7 - 1;  // 7 days.
+  }
+
+  return result;
+}
+
+static bool validate(const Options* opt) {
+  bool valid = true;
+  if (opt->url == NULL) {
+    fprintf(stderr, "error: -url must not be empty.\n");
+    valid = false;
+  }
+  if (opt->cert_url == NULL) {
+    fprintf(stderr, "error: -certUrl must not be empty.\n");
+    valid = false;
+  }
+  if (opt->validity_url == NULL) {
+    fprintf(stderr, "error: -validityUrl must not be empty.\n");
+    valid = false;
+  }
+  if (opt->content == NULL) {
+    fprintf(stderr, "error: -content file must be specified.\n");
+    valid = false;
+  } else {
+    if (access(opt->content, F_OK)) {
+      fprintf(stderr, "error: Cannot access content file: %s\n", opt->content);
+      valid = false;
+    }
+  }
+  if (opt->certificate == NULL) {
+    if (opt->public_key == NULL) {
+      fprintf(
+          stderr,
+          "error: -certificate file or -publicKey file must be specified.\n");
+      valid = false;
+    } else if (access(opt->public_key, F_OK)) {
+      fprintf(stderr,
+              "error: Cannot access publicKey file: %s\n",
+              opt->public_key);
+      valid = false;
+    }
+  } else {
+    if (opt->public_key != NULL) {
+      fprintf(stderr,
+              "error: -certificate and -publicKey cannot be specified "
```

Comment:
Maybe `"error: -certificate file and -publicKey file cannot be specified "` because you write `file` in all other messages.

Having said that, I'd rather suggest removing the optarg (`file`) from all other error messages. It looks verbose (at least to me); the error messages for `-url` etc. don't have the optarg name.